pkg/reconciler/pipelinerun/pipelinerun.go

```diff
@@ -176,6 +176,18 @@ func (c *Reconciler) ReconcileKind(ctx context.Context, pr *v1beta1.PipelineRun)
 	// Read the initial condition
 	before := pr.Status.GetCondition(apis.ConditionSucceeded)
 
+	// Check if we are failing to mark this as timed out for a while. If we are, mark immediately and finish the
+	// reconcile
+	if pr.HasTimedOutForALongTime(ctx, c.Clock) && !isPipelineRunTimeoutConditionSet(pr) {
+		setPipelineRunTimeoutCondition(ctx, pr)
+
+		if err := c.finishReconcileUpdateEmitEvents(ctx, pr, before, nil); err != nil {
+			return err
+		}
+
+		return nil
```

Comment:
Rather than returning `nil` I think we could return a "PermanentError" instead, which means that the key won't be re-queued to the controller but we'll still track this in an error situation and we can provide a context in the error about what is going on - i.e. that the pipeline run condition is still not updated after  2*timeout so we apply a different logic here.

Reply:
👍  I'll change it, but should we consider returning a `controller.NewRequeueImmediately()`? Since we are testing for `!isPipelineRunTimeoutConditionSet(pr)`, it would not be in an infinite loop.